data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/poison/poison_ability.cfg

```diff
@@ -0,0 +1,106 @@
+#textdomain wesnoth-test
+
+#####
+# API(s) being tested: [poison]
+##
+# Actions:
+# Give the leaders a poison ability
+# Have side 1's leader attack side 2's leader with their first weapon
+# Unpoison side 2's leader
+# Have side 1's leader attack side 2's leader with their second weapon
+##
+# Expected end state:
+# Both weapons poison side 2's leader
+#####
+{COMMON_KEEP_A_B_UNIT_TEST "poison_ability" (
+    [event]
+        name = start
+
+        [modify_unit]
+            [filter]
+            [/filter]
+            [effect]
+                apply_to = new_ability
+                [abilities]
+                    {TEST_ABILITY_NO_VALUE poison () SELF=yes}
```

Comment:
Most common use would leave affect_self unset. Not sure if {TEST_ABILITY_NO_VALUE poison () SELF=} is valid, but even that is not exactly same as unset.

Reply:
All such attributes are explicitly set as yes or no, via the default value of the arguments.

Reply:
There are _three_ possibilities for `affect_allies` though – yes, no, and unset have 3 different meanings.

Reply:
For the context of these tests, the `SELF` argument is defined as `no` by default. None of the tests I've added for abilities checks whether unset means yes or no for these attributes.

Reply:
It looks like you're not testing `affect_allies` at all, so it's fine. I'm just noting that that approach won't work if you add `affect_allies` tests later.

Reply:
There's `poison_affect_allies`?

Reply:
Well, poison in and of itself is a weapon special, which doesn't support that. But I'm pretty sure it can be used leadership-style as an ability, which _would_ support `affect_allies`.

Also, my comment wasn't intended to be specific to poison in the first place. It was a general comment on ability tests.

Reply:
I'm lost - there *are* tests in this PR which use poison as a leadership-style ability. All the ability tests I've added have tests like that.

Reply:
Okay, I looked more closely and found you _are_ testing `affect_allies`, and your tests _are_ missing a case. I assume that's also the case for all the abilities you've already added tests for. There is no test for the case where `affect_allies` is unset, which is different from the case of `affect_allies=no` and also different from the case of `affect_allies=false`.

Reply:
Correct, as I said above:
> None of the tests I've added for abilities checks whether unset means yes or no for these attributes.

Reply:
For `affect_allies` the answer is "neither". So you're missing an entire branch of testing for `affect_allies`.

Reply:
That doesn't make sense. It's a boolean - either it affects allies or it doesn't.

Reply:
The 3-way distinction I'm describing is clearly documented on the wiki. You can can argue about whether it's a good way to specify it (maybe you think it shouldn't be a boolean, or maybe you think it should be split into two keys, or whatever), but that isn't really related to unit testing.

Reply:
...I suppose it does. What an absolutely bizarre way to have it work, especially given it doesn't do anything without `affect_adjacent` which has a SUF anyway.

Reply:
Is that actually true? It does say that on the wiki, I'm just wondering whether it's actually accurate on that point.

If it were up to me, I'd say setting `affect_allies` or `affect_enemies` to anything other than "no" without having an `[affect_adjacent]` tag should be treated as equivalent to having an empty `[affect_adjacent]` tag.

But I do see your point that `[affect_adjacent]` pretty much renders the other two obsolete, as far as I can tell.

Reply:
It's true - it's why all the tests for `affect_allies` and `affect_enemies` have the empty `[affect_adjacent]` tags.

Reply:
if affact_allies not specified, abilitie affect same side only if affect_allies=yes afect all unit allied with own(ownside and allied sides)